src/commissioner/Discoverer.cpp

```diff
@@ -0,0 +1,264 @@
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR DnssdDiscoverer::GetAddress(chip::Optional<Transport::PeerAddress> & address,
+                                       chip::Optional<ReliableMessageProtocolConfig> & mrpConfig)
+{
+    if (mNodeIdx >= GetDiscoveredNodes().size() || mIpIdx >= GetDiscoveredNodes().data()[mNodeIdx].numIPs)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    Dnssd::DiscoveredNodeData & node = GetDiscoveredNodes().data()[mNodeIdx];
+    Inet::InterfaceId interfaceId = node.ipAddress[mIpIdx].IsIPv6LinkLocal() ? node.interfaceId[mIpIdx] : Inet::InterfaceId::Null();
+#if 0
+        // TODO(#13326): connecting via TCP does not work
+        if (node.supportsTcp)
+        {
+            address.SetValue(Transport::PeerAddress::TCP(node.ipAddress[mIpIdx], node.port, interfaceId));
+            mrpConfig.SetValue(node.GetMRPConfig());
+        }
+        else
+#endif
+    {
+        address.SetValue(Transport::PeerAddress::UDP(node.ipAddress[mIpIdx], node.port, interfaceId));
+        mrpConfig.SetValue(node.GetMRPConfig());
+    }
+    return CHIP_NO_ERROR;
+}
+
+#endif // CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+
+Discoverer::Discoverer(Controller::DeviceControllerSystemState & systemState, Platform::SharedPtr<SetupPayload> payload,
+                       Delegate * delegate) :
+    Joinable(&mDelegate),
+    mDelegate(delegate), mPayload(payload)
+#if CONFIG_NETWORK_LAYER_BLE
+    ,
+    mBleDiscoverer(systemState, &mDelegate)
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    ,
+    mDnssdDiscoverer(systemState, &mDelegate)
+#endif
+{}
+
+CHIP_ERROR Discoverer::Init()
+{
+    RetainShutdownToken(Platform::MakeShared<ShutdownToken>(&mDelegate));
+    return InProgress() ? CHIP_NO_ERROR : CHIP_ERROR_NO_MEMORY;
+}
+
+void Discoverer::Shutdown()
+{
+#if CONFIG_NETWORK_LAYER_BLE
+    mBleDiscoverer.Shutdown();
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    mDnssdDiscoverer.Shutdown();
+#endif
+    ReleaseShutdownToken();
+}
+
+void Discoverer::SetDelegate(Delegate * delegate)
+{
+    mDelegate = delegate;
+}
+
+CHIP_ERROR Discoverer::Discover()
+{
+    CHIP_ERROR err = CHIP_ERROR_NOT_IMPLEMENTED;
+#if CONFIG_NETWORK_LAYER_BLE
+    bool searchAllOver = this->mPayload.get()->rendezvousInformation == RendezvousInformationFlag::kNone;
+    if (searchAllOver || this->mPayload.get()->rendezvousInformation == RendezvousInformationFlag::kBLE)
+    {
+        SuccessOrExit(err = mBleDiscoverer.StartBleDiscovery(*mPayload.get(), GetShutdownToken()));
+    }
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    // We always want to search on-network because any node that has
+    // already been commissioned will use on-network regardless of
+    // onboarding payload contents.
+    SuccessOrExit(err = mDnssdDiscoverer.StartDnssdDiscovery(*mPayload.get(), GetShutdownToken()));
+#endif
+exit:
+    return err;
+}
+
+CHIP_ERROR Discoverer::GetNextCandidate(Commissionee & commissionee)
```

Comment:
I'm wondering if this is actually the right place for this. We've been updating the device proxy so it has access to all the dnssd address records. If this implementation was internal to the device proxy, it could be used for CASE as well, and we could centralize the logic for this.

Reply:
The problem that I see with combining these is that commissionable node discovery and operational discovery are actually sufficiently different that I have not found that combining these works well.

Commissionable node discovery is a browse-type operation, possibly conducted over multiple network fabrics, where multiple candidates may be located.

In contrast, operational discovery is a resolve-type operation, always by DNS-SD, and can be reasonably approximated as something where only one possible candidate is exists (this was part of the point of the compressed fabric ID, which allows for such an approximation: if fabrics are properly administered, the chance of colliding operational IDs on the same network is vanishingly small).

Reply:
Yes, this is doing both.  It is iterating through both candidates and candidate addresses.

This PR separately iterates through addresses just of the first and only resolved node for CASE here: https://github.com/msandstedt/connectedhomeip/blob/85338f7ac9d4de73dd92406b1b8a66ca9ba3af89/src/commissioner/States.h#L741